pkg/bgpv1/manager/reconciler/service.go

```diff
@@ -298,31 +293,61 @@ func (r *LBServiceReconciler) svcDesiredRoutes(newc *v2alpha1api.CiliumBGPVirtua
 	if !svcSelector.Matches(serviceLabelSet(svc)) {
 		return nil, nil
 	}
-
-	// Ignore service managed by an unsupported LB class.
-	if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != v2alpha1api.BGPLoadBalancerClass {
-		// The service is managed by a different LB class.
-		return nil, nil
-	}
-
 	// Ignore externalTrafficPolicy == Local && no local endpoints.
 	if svc.Spec.ExternalTrafficPolicy == slim_corev1.ServiceExternalTrafficPolicyLocal &&
 		!hasLocalEndpoints(svc, ls) {
 		return nil, nil
 	}
-
 	var desiredRoutes []netip.Prefix
-	for _, ingress := range svc.Status.LoadBalancer.Ingress {
-		if ingress.IP == "" {
-			continue
-		}
-
-		addr, err := netip.ParseAddr(ingress.IP)
-		if err != nil {
-			continue
+	// Loop over the service advertisements and determine the desired routes.
+	for _, svcAdv := range newc.ServiceAdvertisements {
+		switch svcAdv {
+		case v2alpha1api.BGPLoadBalancerIPAddr:
+			if svc.Spec.Type != slim_corev1.ServiceTypeLoadBalancer {
+				continue
+			}
+			// Ignore service managed by an unsupported LB class.
+			if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != v2alpha1api.BGPLoadBalancerClass {
+				// The service is managed by a different LB class.
+				return nil, nil
+			}
+			for _, ingress := range svc.Status.LoadBalancer.Ingress {
+				if ingress.IP == "" {
+					continue
+				}
+				addr, err := netip.ParseAddr(ingress.IP)
+				if err != nil {
+					continue
+				}
+				desiredRoutes = append(desiredRoutes, netip.PrefixFrom(addr, addr.BitLen()))
+			}
+		case v2alpha1api.BGPClusterIPAddr:
+			if svc.Spec.Type != slim_corev1.ServiceTypeClusterIP {
+				continue
+			}
+			if svc.Spec.ClusterIP == "" || len(svc.Spec.ClusterIPs) == 0 {
+				return nil, nil
+			}
+			if svc.Spec.ClusterIP != "" {
```

Comment:
Note that `ClusterIP` / `ClusterIPs` can have a valid value "None", which will always fail address parsing, so it is probably worth skipping it explicitly: https://github.com/cilium/cilium/blob/4082bc38407a3fd2b59f2bd1e8e0e1ed06be8340/vendor/k8s.io/api/core/v1/types.go#L5264-L5266

Reply:
Thank you for the reminder.